### DISCUSSION TITLE
IPv6 & TrueNAS SCALE NFS service

### DISCUSSION THREAD

**Discussion by Miky028** [ipxe/ipxe#1098] [Q&A]
Hello,
I'm trying to run iPXE in dual-stack, but I have a problem with IPv6.

Connecting with IPv6 to the TrueNAS SCALE 23.10.0.1 (Cobia) NFS share ended up with error: Connection reset.
Connecting with IPv6 to the TrueNAS CORE NFS share works fine.

Connecting with IPv4 to the TrueNAS SCALE 23.10.0.1 (Cobia) NFS share works fine.
Connecting with IPv4 to the TrueNAS CORE NFS share works fine.

Packet capture log of the IPv6 connection:

client server TCP 98 833 → 111 [SYN] Seq=0 Win=65532 Len=0 TSval=6848 TSecr=0 SACK_PERM WS=512 MSS=1440
server client TCP 94 111 → 833 [SYN, ACK] Seq=0 Ack=1 Win=64260 Len=0 MSS=1440 SACK_PERM TSval=3776116643 TSecr=6848 WS=128
client server Portmap 170 V2 GETPORT Call (Reply In 2302) MOUNT(100005) V:3 TCP
server client TCP 86 111 → 833 [ACK] Seq=1 Ack=85 Win=64256 Len=0 TSval=3776116644 TSecr=6848
server client Portmap 118 V2 GETPORT Reply (Call In 2298) Port:43573
client server Portmap 170 V2 GETPORT Call (Reply In 2318) NFS(100003) V:3 TCP
server client Portmap 118 V2 GETPORT Reply (Call In 2310) Port:2049
client server TCP 86 833 → 111 [FIN, ACK] Seq=169 Ack=65 Win=2097152 Len=0 TSval=6848 TSecr=3776116644
server client TCP 86 111 → 833 [FIN, ACK] Seq=65 Ack=170 Win=64256 Len=0 TSval=3776116644 TSecr=6848
client server TCP 86 833 → 111 [ACK] Seq=170 Ack=66 Win=2097152 Len=0 TSval=6848 TSecr=3776116644
client server TCP 98 477 → 43573 [SYN] Seq=0 Win=65532 Len=0 TSval=6848 TSecr=0 SACK_PERM WS=512 MSS=1440
server client TCP 74 43573 → 477 [RST, ACK] Seq=1 Ack=1 Win=0 Len=0

iPXE DEBUG log:

TCP 0x5d7699a8 TX 833->111 71a08530..71a08531 00000000 0 SYN
TCP 0x5d7699a8 RX 833<-111 71a08531 2e2fd012..2e2fd013 0 SYN ACK
TCP 0x5d7699a8 using timestamps, SACK, TX window x128, RX window x512
TCP 0x5d7699a8 transitioned from SYN_SENT to ESTABLISHED
NFS_OPEN 0x5d76aae8 GETPORT call (mount)
TCP 0x5d7699a8 TX 833->111 71a08531..71a08585 2e2fd013 84 PSH ACK
TCP 0x5d7699a8 RX 833<-111 71a08585 2e2fd013..2e2fd013 0 ACK
TCP 0x5d7699a8 RX 833<-111 71a08585 2e2fd013..2e2fd033 32 PSH ACK
NFS_OPEN 0x5d76aae8 got GETPORT reply (mount)
NFS_OPEN 0x5d76aae8 GETPORT call (nfs)
TCP 0x5d7699a8 TX 833->111 71a08585..71a085d9 2e2fd033 84 PSH ACK
TCP 0x5d7699a8 RX 833<-111 71a085d9 2e2fd033..2e2fd053 32 PSH ACK
NFS_OPEN 0x5d76aae8 got GETPORT reply (nfs)
TCP 0x5d7699a8 transitioned from ESTABLISHED to FIN_WAIT_1
TCP 0x5d7699a8 TX 833->111 71a085d9..71a085da 2e2fd053 0 FIN ACK
TCP 0x5d7699a8 RX 833<-111 71a085da 2e2fd053..2e2fd054 0 FIN ACK
TCP 0x5d7699a8 transitioned from FIN_WAIT_1 to TIME_WAIT
TCP 0x5d7699a8 TX 833->111 71a085da..71a085da 2e2fd054 0 ACK
TCP 0x5d769e08 allocated
TCP 0x5d769e08 transitioned from CLOSED to SYN_SENT
TCP 0x5d769e08 bound to port 477
TCP 0x5d769e08 timer fired in SYN_SENT for 6b6fab61..6b6fab61 00000000
TCP 0x5d769e08 TX 477->43573 6b6fab61..6b6fab62 00000000 0 SYN
TCP 0x5d769e08 RX 477<-43573 6b6fab62 00000000..00000000 0 RST ACK
TCP 0x5d769e08 transitioned from SYN_SENT to CLOSED
NFS_OPEN 0x5d76aae8 completed (Connection reset (https://ipxe.org/0f0a6095))
NFS_OPEN 0x5d76aae8 freed
TCP 0x5d769e08 connection deleted
TCP 0x5d769e08 connection reset by peer
IPv6 received packet rejected by stack: Connection reset (https://ipxe.org/0f0a6095)
Connection reset (https://ipxe.org/0f0a6095)

It looks like that iPXE is trying connect to NFS server using version 3. How can I force to use version 4? Maybe it would be succesfull.
I have no problem with TrueNAS SCALE NFS share with other apps.
Where can be a problem?
THX for advice.

Miky

**Comment by NiKiZe** [comment]
Currently iPXE only has v3 support https://github.com/ipxe/ipxe/blob/f22879ca994aedd2667bb0c73ebce505bf5f8cef/src/include/ipxe/nfs.h#L19

**Comment by Miky028** [comment]
OK thx for replay, so NFSv4 is no go and have to crack current problem than.
My previous comment may imply that I have configured only NFSv4 share. It support both versions.
